sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala

```diff
@@ -238,7 +238,12 @@ case class CachedRDDBuilder(
   }
 
   def isCachedColumnBuffersLoaded: Boolean = {
-    _cachedColumnBuffers != null && isCachedRDDLoaded
+    if (_cachedColumnBuffers != null) {
+      _cachedColumnBuffers.synchronized {
+        return _cachedColumnBuffers != null && isCachedRDDLoaded
+      }
+    }
+    false
   }
 
   def isCachedRDDLoaded: Boolean = {
```

Comment:
This change is really helpful, is it a good idea to add another synchronized block in `isCacheRDDLoaded` too? It seems `_cachedColumnBuffersAreLoaded` is also volatile and while there is no other call to `isCacheRDDLoaded`, adding that should be safe for another method to call it.

Reply:
> Don't we just need a similar synchronized block?

You are right, updated.

Reply:
> This change is really helpful, is it a good idea to add another synchronized block in `isCacheRDDLoaded` too? It seems `_cachedColumnBuffersAreLoaded` is also volatile and while there is no other call to `isCacheRDDLoaded`, adding that should be safe for another method to call it.

Added, anyway adding a synchronized block is not a bad idea.